extensions/amp-image-viewer/0.1/amp-image-viewer.js

```diff
@@ -261,104 +245,121 @@ export class AmpImageViewer extends AMP.BaseElement {
   }
 
   /**
-   * @return {number}
+   * Sets the source width and height based on the user-defined
+   * dimensions of the amp-img if exists, the natural dimensions
+   * of the source image if loaded, and the offset dimensions of
+   * amp-img element if not.
+   * @param {!Element}
    * @private
    */
-  getSourceWidth_() {
-    if (this.sourceAmpImage_.hasAttribute('width')) {
-      return parseInt(this.sourceAmpImage_.getAttribute('width'), 10);
+  setSourceDimensions_(ampImg) {
+    let width, height;
+    if ((width = ampImg.getAttribute('width'))
+        && (height = ampImg.getAttribute('height'))) {
+      this.sourceWidth_ = width;
+      this.sourceHeight_ = height;
     } else {
-      const img = elementByTag(dev().assertElement(this.sourceAmpImage_),
-          'img');
-      return img ? img.naturalWidth : this.sourceAmpImage_./*OK*/offsetWidth;
-    }
-  }
-
-  /**
-   * @return {number}
-   * @private
-   */
-  getSourceHeight_() {
-    if (this.sourceAmpImage_.hasAttribute('height')) {
-      return parseInt(this.sourceAmpImage_.getAttribute('height'), 10);
-    } else {
-      const img = elementByTag(dev().assertElement(this.sourceAmpImage_),
-          'img');
-      return img ? img.naturalHeight : this.sourceAmpImage_./*OK*/offsetHeight;
+      const img = elementByTag(ampImg, 'img');
+      if (img && (width = img.naturalWidth) && (height = img.naturalHeight)) {
+        this.sourceWidth_ = width;
+        this.sourceHeight_ = height;
+      } else {
+        this.sourceWidth_ = ampImg./*OK*/offsetWidth;
+        this.sourceHeight_ = ampImg./*OK*/offsetHeight;
+      }
     }
   }
 
   /**
    * Initializes the image viewer to the target image element such as
    * "amp-img". The target image element may or may not yet have the img
    * element initialized.
+   * @returns {!Promise}
    */
   init_() {
-    this.sourceWidth_ = this.getSourceWidth_();
-    this.sourceHeight_ = this.getSourceHeight_();
-    this.srcset_ = srcsetFromElement(dev().assertElement(this.sourceAmpImage_));
+    if (this.image_) {
+      return Promise.resolve();
+    }
 
-    ARIA_ATTRIBUTES.forEach(key => {
-      if (this.sourceAmpImage_.hasAttribute(key)) {
-        this.image_.setAttribute(key, this.sourceAmpImage_.getAttribute(key));
-      }
-    });
-    st.setStyles(dev().assertElement(this.image_), {
-      top: st.px(0),
-      left: st.px(0),
-      width: st.px(0),
-      height: st.px(0),
+    this.image_ = this.element.ownerDocument.createElement('img');
+    this.image_.classList.add('i-amphtml-image-viewer-image');
+    const ampImg = dev().assertElement(this.sourceAmpImage_);
+    this.setSourceDimensions_(ampImg);
+    this.srcset_ = srcsetFromElement(ampImg);
+
+    return this.mutateElement(() => {
+      ARIA_ATTRIBUTES.forEach(key => {
+        let k;
+        if ((k = this.sourceAmpImage_.getAttribute(key))) {
```

Comment:
nit: prefer to assign variables outside of `if` conditions

Reply:
Okay: 

```
ARIA_ATTRIBUTES.forEach(key => {
  if (this.sourceAmpImage_.hasAttribute(key)) {
    this.image_.setAttribute(key, this.sourceAmpImage_.getAttribute(key));
  }
});
```
163 bytes, 166 bytes minified

```
ARIA_ATTRIBUTES.forEach(key => {
  let k;
  if ((k = this.sourceAmpImage_.getAttribute(key))) {
    this.image_.setAttribute(key, k);
  }
});
```
141 bytes, 137 bytes minified

Since the logic here is pretty much: set if exists, I think we can leave this as is.